user/TASK/chassis_task/chassis_task.c

```diff
@@ -250,6 +269,91 @@ static void increment_PID(Chassis_t *chassis_pid){
         chassis_pid->motor[BACK_LEFT].speed_read, 
         chassis_pid->motor[BACK_LEFT].current_out);
         serial_send_string(pid_out);
+        
     }
-	
+}
+
+
+
+
+/**
+* Current limiter has 4 states,
+* If over move to half
+* If over still for longer than hysteresis time, move to quarter,
+* If still over for longer than hysteresis time, move to off.
+* for all of these cases:
+* If current stays below limit for hysteresis time, reduce limitation by 1 level. 
+*/
+
+static void check_allowed_current(Chassis_t *chassis_feedback){
+    for(int i = 0; i < 4; i++){
+       if(abs(chassis_feedback->motor[i].motor_feedback->current_read) > CURRENT_LIMIT){
+           sprintf(message, "current limit reached, limiter state is %d", chassis_feedback->motor[i].limiter);
+           serial_send_string(message);
+            if(chassis_feedback->motor[i].limiter == FULL_CURRENT){
+                chassis_feedback->motor[i].limiter = HALF_CURRENT;
+            }
+            
+            chassis_feedback->motor[i].limiter_counter++;
+            if(chassis_feedback->motor[i].limiter_counter > HYSTERESIS_PERIOD){
+                if(chassis_feedback->motor[i].limiter != NO_CURRENT){
+                    chassis_feedback->motor[i].limiter++;
+                }
+                
+                chassis_feedback->motor[i].limiter_counter = 0;                
+            }
+        }
+        else{
+            chassis_feedback->motor[i].limiter_counter--;
+            if(chassis_feedback->motor[i].limiter_counter < -HYSTERESIS_PERIOD){
+                if(chassis_feedback->motor[i].limiter != FULL_CURRENT){
+                    chassis_feedback->motor[i].limiter--;
+                }
+                
+                chassis_feedback->motor[i].limiter_counter = 0;
+            }
+        }
+        
+        limit_current(&chassis_feedback->motor[i]);
+    }
+}
+
+    
+static void limit_current(Chassis_Motor_t *motor){
+    sprintf(message, "limiter state is changing from %d", motor->limiter);
+    switch(motor->limiter){
+        case FULL_CURRENT:
+            break;
+        case HALF_CURRENT:
+            motor->current_out *= 0.5f;
+            break;
+        case QUARTER_CURRENT:
+            motor->current_out *= 0.25f;
+            break;
+        case NO_CURRENT:
+            motor->current_out = 0;
+            break;
+    }
+    
+}
+
+//static void under_current_limit(Chassis_Motor_t motor
+    
+
+    
+
+
+
```

Comment:
What is this egregious whitespace doing here Benjamin _Brown_! Some more whitespace between increment_PID() and check_allowed_current(). We should call you Benjamin _White_